java/arcs/core/data/proto/policy.proto

```diff
@@ -0,0 +1,107 @@
+syntax = "proto3";
+
+package arcs;
+
+option java_package = "arcs.core.data.proto";
+option java_multiple_files = true;
+
+// Defines a data usage policy.
+message PolicyProto {
+  // Policy name.
+  string name = 1;
+
+  // Targets governed by the policy.
+  repeated PolicyTargetProto targets = 2;
+
+  // Additional config metadata for the policy.
+  repeated PolicyConfigProto configs = 3;
+
+  // Human readable description for the policy.
+  string description = 4;
+
+  enum EgressType {
+    EGRESS_TYPE_UNSPECIFIED = 0;
+    LOGGING = 1;
+    FEDERATED_ANALYTICS = 2;
+  }
+  // Egress type permitted by the policy.
+  EgressType egress_type = 5;
+
+  // Whether user consent is required before using the policy.
+  bool user_consent_required = 6;
+
+  enum ApprovalCondition {
+    APPROVAL_CONDITION_UNSPECIFIED = 0;
+    PRE_APPROVED = 1;
+    APPROVAL_REQUIRED = 2;
+  }
+  // Approval condition for the policy.
+  ApprovalCondition approval_condition = 7;
+}
+
+message PolicyTargetProto {
+  // ID of the store targeted by the policy.
+  string store_id = 1;
+
+  // Valid retention options for storing the target data.
+  repeated PolicyRetentionProto retentions = 2;
+
+  // Allowed usages for all fields in this target type.
+  repeated PolicyFieldProto fields = 3;
+}
+
+message PolicyFieldProto {
+  // Field name.
+  string name = 1;
+
+  enum UsageType {
+    USAGE_TYPE_UNSPECIFIED = 0;
+    // Can be used for anything.
+    ANY = 1;
+    // Can be egressed.
+    EGRESS = 2;
+    // Can be used in joins, but not read.
+    JOIN = 3;
+  }
+
+  message AllowedUsages {
+    // Indicates all types of usages that are permitted. An empty list indicates
+    // no usages are permitted.
+    repeated UsageType usages = 1;
+  }
+
+  // Specifies the allowed usages of raw data.
+  AllowedUsages raw_usages = 2;
+
+  // Specifies the allowed usages given that certain redactions have been
+  // performed first. Maps from redaction label to AllowedUsages.
+  map<string, AllowedUsages> redacted_usages = 3;
+
+  // Specified usages for subfields nested inside this field.
+  repeated PolicyFieldProto subfields = 4;
```

Comment:
How do we handle tuples here?

E.g. with a tuple of references:
```
Product {
  name: Text.
  reviews: [(&Author {nickname: Text}, &Review {content: Text, rating: Number})]
}
```

Myk recently added tuple fields to Kotlin, maybe they don't work quite yet, but they will probably pop up at some point.

I want to write a proposal for expressions that will allow targeting fields in schemas (while navigating references, tuples etc), something like XPath for Schemas, which would standardize how we do this across references, type adapters and a few other places (e.g. `join (user.friends, twitterPersonas)`). That could be one avenue to explore.

Reply:
XPaths for schemas would be a great idea! I added some code in https://github.com/PolymerLabs/arcs/pull/5357 to validate basic dotted field paths against a schema, but as you say, it doesn't handle tuples. It would be great to standardise that syntax and write some comprehensive code supporting it.